scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+4. `scanner.v4.Matcher/GetVulnerabilityMetadata`: Get information on vulnerability metadata, e.g., last update timestamp.
+
+Example of manifest's `hash_id` usage to create index reports for container images:
+
+```
+message ContainerImageLocator {
+    string url      = 1;
+    string username = 2;
+    string password = 3;
+}
+
+message CreateIndexRequest {
+    string hash_id = 1;
+    oneof resource_locator {
+        ContainerImageLocator container_image;
+    }
+}
+```
+
+Example of index request and vulnerability request:
+
+```
+message GetIndexRequest {
+    string hash_id   = 1;
+    // If true, does not return the Index Report, only status.
+    bool check_only  = 2;
+}
+
+message GetVulnerabilitiesRequest {
+    string hash_id = 1;
+}
+```
+
+Examples of common error codes and scenarios:
+
+| Code | Scenario |
+| --- | ---
+| `UNAVAILABLE` | Scanner temporary or retriable issue caused by dependencies (e.g., registries, databases).
+| `INVALID_ARGUMENT` | Invalid input or malformed input (e.g., bad credentials, malformed `hash_id`).
+| `INTERNAL` | Unrecoverable errors, unexpected conditions, or invariant (e.g., bugs, broken schema, etc).
+| `FAILED_PRECONDITION	` | Index report does not exist for this given `hash_id` (e.g. when calling `GetVulnerabilities`).
+
+All APIs are idempotent, hence retriable on timeouts and temporary errors.
+
+## Consequences
+
+Leveraging gRPC-only helps maintain service contracts and enhances performance. It also aligns with other Stackrox services.
+
+The new image integration for ScannerV4 ensures minimal changes are necessary for Central to switch between ScannerV2 and ScannerV4 using a feature flag.
+
+The Indexer and Matcher modes allow additional deployment options, enabling ScannerV4 to scale and serve different scenarios. The multiple modes and gRPC services pave the way for advanced deployments, such as multi-tenant scanning and node scanning.
+
+Using gRPC exclusively increases complexity for load balancing ScannerV4 for horizontal scalability. Additional steps will be necessary to achieve this, including service-mesh tooling or client-side load-balancing, which are not available out of the box and require additional work.
+
+The client-managed `hash_id` approach for identifying manifests may lead to duplicated index reports for locators pointing to the same resource. While this simplifies the API to support future resource types (orchestrator, nodes), it may lead to client misuse.
+
+Always re-indexing upon `Indexer/CreateIndex` calls is sub-optimal. There are [interfaces in ClairCore's Indexer`](https://github.com/quay/clair/blob/8174e950186c03bee10a9174643bca0f173710c2/indexer/service.go#L47) that allows check to not trigger re-indexing. This could potentially be leveraged by ScannerV4 to optimize re-indexing, a transparent change for customers.
```

Comment:
Because layers are content-addressed, re-indexing on every request is just running CPUs for fun.

Reply:
Unless Scanner API stores and compares the indexer state, or deletes IndexReports periodically, etc., we will have to allow re-indexing to updates to Scanner would be reflected in the stored IndexerReports. 

That would only happen during re-scans (periodically current in Central). We will not do this before requesting vulnerabilities because Matcher will always do gets. Does that make sense to you?

Reply:
Clair handles this by allowing conditional requests with a "state token" that represents the current state of the indexer.

Reply:
Why would we reindex upon every call? We can check if there is already an index report for the image before asking to index.

Though something to think about: how do we determine if we need to reindex an image if a package scanner in ClairCore has updated?

Reply:
> Clair handles this by allowing conditional requests with a "state token" that represents the current state of the indexer.

We would have to do the same to avoid re-indexing, passing the state token to get conditional updates.